drivers/ieee802154/ieee802154_kw41z.c

```diff
@@ -126,67 +263,153 @@ static inline void kw41z_wait_for_idle(void)
+		/* wait for Sequence Idle (if not already) */
+
+		while (ZLL->SEQ_STATE & ZLL_SEQ_STATE_SEQ_STATE_MASK) {
+		}
 	}
 
-	kw41z_wait_for_idle();
+	/* Stop timers */
+	ZLL->PHY_CTRL &= ~(ZLL_PHY_CTRL_TMR1CMP_EN_MASK |
+			ZLL_PHY_CTRL_TMR2CMP_EN_MASK |
+			ZLL_PHY_CTRL_TMR3CMP_EN_MASK |
+			ZLL_PHY_CTRL_TC3TMOUT_MASK);
 
-	return 0;
-}
+	/* clear all IRQ bits to avoid unexpected interrupts */
+	ZLL->IRQSTS &= -1;
 
-static inline void kw41z_enable_seq_irq(void)
-{
-	ZLL->PHY_CTRL &= ~ZLL_PHY_CTRL_SEQMSK_MASK;
+	irq_unlock(key);
 }
 
-static inline void kw41z_disable_seq_irq(void)
+static void kw41z_isr_timeout_cleanup(void)
 {
-	ZLL->PHY_CTRL |= ZLL_PHY_CTRL_SEQMSK_MASK;
+	u32_t irqsts;
+
+	/*
+	 * Set the PHY sequencer back to IDLE and disable TMR3 comparator
+	 * and timeout
+	 */
+	ZLL->PHY_CTRL &= ~(ZLL_PHY_CTRL_TMR3CMP_EN_MASK |
+			ZLL_PHY_CTRL_TC3TMOUT_MASK   |
+			ZLL_PHY_CTRL_XCVSEQ_MASK);
+
+	/* Mask SEQ, RX, TX and CCA interrupts */
+	ZLL->PHY_CTRL |= ZLL_PHY_CTRL_CCAMSK_MASK |
+			ZLL_PHY_CTRL_RXMSK_MASK  |
+			ZLL_PHY_CTRL_TXMSK_MASK  |
+			ZLL_PHY_CTRL_SEQMSK_MASK;
+
+	while (ZLL->SEQ_STATE & ZLL_SEQ_STATE_SEQ_STATE_MASK) {
+	}
+
+	irqsts = ZLL->IRQSTS;
+	/* Mask TMR3 interrupt */
+	irqsts |= ZLL_IRQSTS_TMR3MSK_MASK;
+
+	ZLL->IRQSTS = irqsts;
 }
 
-static void kw41z_tmr1_set_timeout(u32_t timeout)
+static void kw41z_isr_seq_cleanup(void)
 {
-	timeout += ZLL->EVENT_TMR >> ZLL_EVENT_TMR_EVENT_TMR_SHIFT;
+	u32_t irqsts;
+
+	/* Set the PHY sequencer back to IDLE */
+	ZLL->PHY_CTRL &= ~ZLL_PHY_CTRL_XCVSEQ_MASK;
+	/* Mask SEQ, RX, TX and CCA interrupts */
+	ZLL->PHY_CTRL |= ZLL_PHY_CTRL_CCAMSK_MASK |
+			ZLL_PHY_CTRL_RXMSK_MASK  |
+			ZLL_PHY_CTRL_TXMSK_MASK  |
+			ZLL_PHY_CTRL_SEQMSK_MASK;
 
-	ZLL->PHY_CTRL &= ~ZLL_PHY_CTRL_TMR1CMP_EN_MASK;
+	while (ZLL->SEQ_STATE & ZLL_SEQ_STATE_SEQ_STATE_MASK) {
+	}
 
-	ZLL->T1CMP = timeout;
-	ZLL->IRQSTS = (ZLL->IRQSTS & ~ZLL_IRQSTS_TMR1MSK_MASK) |
-		      ZLL_IRQSTS_TMR1IRQ_MASK;
+	irqsts = ZLL->IRQSTS;
+	/* Mask TMR3 interrupt */
+	irqsts |= ZLL_IRQSTS_TMR3MSK_MASK;
 
-	ZLL->PHY_CTRL |= ZLL_PHY_CTRL_TMR1CMP_EN_MASK;
+	/* Clear transceiver interrupts except TMRxIRQ */
+	irqsts &= ~(ZLL_IRQSTS_TMR1IRQ_MASK |
+		ZLL_IRQSTS_TMR2IRQ_MASK |
+		ZLL_IRQSTS_TMR3IRQ_MASK |
+		ZLL_IRQSTS_TMR4IRQ_MASK);
+	ZLL->IRQSTS = irqsts;
 }
 
-static inline void kw41z_tmr1_disable(void)
+static inline void kw41z_enable_seq_irq(void)
 {
-	ZLL->IRQSTS |= ZLL_IRQSTS_TMR1MSK_MASK;
-	ZLL->PHY_CTRL &= ~ZLL_PHY_CTRL_TMR1CMP_EN_MASK;
+	ZLL->PHY_CTRL &= ~ZLL_PHY_CTRL_SEQMSK_MASK;
 }
 
-static void kw41z_tmr2_set_timeout(u32_t timeout)
+static inline void kw41z_disable_seq_irq(void)
 {
-	timeout += ZLL->EVENT_TMR >> ZLL_EVENT_TMR_EVENT_TMR_SHIFT;
+	ZLL->PHY_CTRL |= ZLL_PHY_CTRL_SEQMSK_MASK;
+}
 
-	ZLL->PHY_CTRL &= ~ZLL_PHY_CTRL_TMR2CMP_EN_MASK;
+/*
+ * Set the T3CMP timer comparator. The 'timeout' value is an offset from
+ * now.
+ */
+static void kw41z_tmr3_set_timeout(u32_t timeout)
+{
+	u32_t irqSts;
```

Comment:
no camel case please. (Only HAL ones can't be changed.) Apply that where relevant.